### DISCUSSION TITLE
distances returned by search and calc_distance for the same vectors and metric are different

### DISCUSSION THREAD

**Discussion by drassokhin** [milvus-io/milvus#19688] [Q&A and General discussion]
If I insert two binary vectors (first_vector and second_vector) into an empty collection (which has a corresponding field called "fingerprint") and run a search for the first vector (without creating any indices),
```
vectors = [first_vector]
search_params = {"metric_type": "TANIMOTO"}
res = collection.search(vectors, anns_field="fingerprint", param=search_params, limit=2)
```
the search results will contain two hits (as expected), the one corresponding to the first vector (with distance==0, which is also expected), and the second one matching the second vector, with the corresponding distance that happens to be 1.0473. Why is it >1.0, and why is it different from the distance calculated via a direct call calc_distance for the same vectors with the same metric (TANIMOTO)? 
```
vectors1 = [first_vector]
vectors2 = [second_vector]
params = {"metric_type": "TANIMOTO"}
utility.calc_distance(vectors1, vectors2, params=params, using='default')
```
returns [[0.939393937587738]]

**Comment by yhmo** [comment]
The Tanimoto coefficient of vector A and B can be defined as:

T = Nc / (Na + Nb - Nc)

- Na is the number of bits in vector А
- Nb is the number of bits in vector B
- Nc is the number of bits that are equal in A and B

If A and B are the same, the Tanimoto coefficient should be 1.0. So, if the top1 is 0.0, I think it is incorrect.

Do you have a script to reproduce this?

**Reply by drassokhin** [reply]
Here is a complete working example:
```python
from pymilvus import connections, FieldSchema, CollectionSchema, DataType, Collection, utility
# this needs the bitstring package
import bitstring
COLLECTION_NAME = 'tanimoto_test'
VECTOR_DIMENSION = 1024
METRIC_TYPE = 'TANIMOTO'
HOST = '<your_milvus_host_address>'
PORT = '<your_milvus_port>'  # has to be a string


V1 = b'\x00\x00\x00\x00\x02\x00\x00\x00\x00\x01\x00\x00\x00\x00\x04\x00\x00\x00\x00\x00\x08\x00\x00\x00\x00\x00'\
     b'\x02\x00\x80\xc2\x00\x00\x00P\x00\x00\x00\x00\x00\x00\x00\x00\x08\x00\x10\x00\x00\x04\x00\x01\x00\x00\x00'\
     b'\x10\x00\x00\x00\x80\x01\x00\x00\x02\x80\x00\x00\x00@\x00@\x00\x03(B\x00\x00\x00\x00\x00\x00\x00\x00\x04h\x02'\
     b'\x01\x00\x80\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x02\x04\x02\x80\x00\x00\x00H\x00\x02\x00\x00\x00\x02(\x00'\
     b'\x00\x00@\x80\x00\x08\x00\x11\x00\x00\x00\x00\x02\x00\x08'

V2 = b'\x00\x00\x00\x00\x02\x00\x00\x00\x00\x08\x00\x00\x00\x00\x00\x00\x00\x00\x00\x01\x00\x00\x04\x00\x00\x00\x02'\
     b'\x00\x00\x02\x00\x00\x00@\x00 \x00\x00\x00\x00\x00\x00\x08\x00\x10\x00\x00\x04\x00\x00\x00\x00\x00\x10\x00'\
     b'\x00@\x80\x01\x00\x00\x02\x00\x00\x00\x00\x00\x00@\x00\x02\x08B\x00\x80\x00\x00\x00\x00\x00\x00\x00)\x00\x01'\
     b'\x00\x00\x00\x00\x00\x04\x00\x00\x00\x00\x00\x10\x02\x04\x02\x80\x00\x00\x00\x00\x00\x02\x00\x00\x00\x00\x00'\
     b'\x02\x00\x00\x00\x80\x00\x08\x00\x11\x00\x00\x00\x00\x02\x00\x08'


def load():
    connections.connect(
        alias='default',
        host=HOST,
        port=PORT)

    # don't try to create collection and load data if it has already been done
    if utility.has_collection(COLLECTION_NAME):
        connections.disconnect('default')
        return

    fld_idx = FieldSchema(name="id", dtype=DataType.INT64, descrition="int64 primary id",
                          is_primary=True)
    fld_fingerprint = FieldSchema(name="fingerprint", dtype=DataType.BINARY_VECTOR,
                                  descrition="binary vector fingerprint",
                                  dim=VECTOR_DIMENSION, is_primary=False)

    schema = CollectionSchema(fields=[fld_idx, fld_fingerprint],
                              description="test")
    collection = Collection(name=COLLECTION_NAME, schema=schema, using='default')
    assert len(V1) == VECTOR_DIMENSION // 8
    assert len(V2) == VECTOR_DIMENSION // 8
    r = collection.insert([[1, 2], [V1, V2]])
    print(f'Inserted: {r}')
    connections.disconnect('default')


def test():
    connections.connect(
        alias='default',
        host=HOST,
        port=PORT)
    collection = Collection(name=COLLECTION_NAME)
    collection.load()
    search_params = {"metric_type": METRIC_TYPE}
    res = collection.search([V1], anns_field="fingerprint", param=search_params, limit=2)
    r = res[0]
    print(f'Search result: {r}')

    v1 = {"bin_vectors": [V1]}
    v2 = {"bin_vectors": [V2]}
    params = {"metric_type": METRIC_TYPE}
    d = utility.calc_distance(v1, v2, params=params)
    print(f'Distance via calc_distance: {d}')
    collection.release()
    connections.disconnect('default')

    ba1 = bitstring.BitArray(auto=V1)
    ba2 = bitstring.BitArray(auto=V2)
    ba_and = ba1 & ba2  #  intersection (common bits)
    ba_or = ba1 | ba2   #  union (all bits)

    print(f'1 bit counts: b1: {ba1.count(1)}, b2: {ba2.count(1)}, intersect: {ba_and.count(1)}, union: {ba_or.count(1)}')

    tanimoto_sim1 = ba_and.count(1) / ba_or.count(1)
    # alternatively, we can compute it like this, which should give us identical results
    tanimoto_sim2 = ba_and.count(1) / (ba1.count(1) + ba2.count(1) - ba_and.count(1))
    print(f'True Tanimoto similarity method 1: {tanimoto_sim1}, method 2: {tanimoto_sim2}')


if __name__ == '__main__':
    load()
    test()
    print('Done.')

```

When this example is executed, it produces the following output:
```
Search result: ['(distance: 0.0, id: 1)', '(distance: 1.047305703163147, id: 2)']
Distance via calc_distance: [[0.939393937587738]]
1 bit counts: b1: 52, b2: 40, intersect: 30, union: 62
True Tanimoto similarity method 1: 0.4838709677419355, method 2: 0.4838709677419355
Done.
```

So, I have no clue how Milvus actually computes the Tanimoto distance between two binary vectors when searching or when calling the compute_distance method, but it is not the same as the 'true' Tanimoto similarity computed directly, neither this is Tanimoto_distance := 1.0 - Tanimoto_similarity, which, for V1 and V2, would be 1.0-0.4838709677419355 = 0.5161290322580645. So, perhaps, Milvus developers would be able to explain us what's going on.

**Reply by yhmo** [reply]
Thank you for pointing out the bug.
It is a shame that the calc_distance() is wrong for Tanimoto.

As we know, Tanimoto coefficient value is T = Nc / (Na + Nb - Nc), but the calc_distance() uses all bits as Na so that it returns an incorrect result:
https://github.com/milvus-io/milvus/blob/a8a074162f448596a1d53c2d0537d5b05956e6e0/internal/util/distance/calc_distance.go#L276
I have created an issue for this bug: https://github.com/milvus-io/milvus/issues/19723 

For the search(), it seems the returned distance is not Tanimoto coefficient. I created another issue for the search engine: https://github.com/milvus-io/knowhere/issues/500

**Comment by cydrain** [comment]
please check https://github.com/milvus-io/knowhere/issues/500

**Reply by drassokhin** [reply]
So, the 'Tanimoto distances' returned by the vector search are correct, but they use the following definition of the Tanimoto distance for bit vectors (or, rather, Tanimoto distance score, because it is not a proper distance metric: ```TanimotoDistanceScore := -log2(JaccardSimilarity)```, as opposed to ```TanimotoDistance := 1.0 - JaccardSimilarity```, which is also commonly used in various software packages and literature. ```JaccardSimilarity(bitsetA, bitsetB) := (bitsetA intersect bitsetB).count(1) / (bitsetA union bitsetB).count(1) ```. If this is the case, the distance returned by the vector search should be -log2(0.4838709677419355), which is, indeed, 1.0473. However, the implementation of utility.calc_distance is incorrect regardless of which definition of Tanimoto distance is used. At any rate, the definitions should be consistent across the entire API and reflected in the documentation.

**Comment by drassokhin** [comment]
I have upgraded Milvus from 2.1.1 to 2.1.4, and the behavior has not changed. calc_distance and search are still using different definitions of the Tanimoto distance between bit vectors and return different values for the same pair of vectors.
